internal/scheduler/schedulerapp.go

```diff
@@ -0,0 +1,146 @@
+	// Database setup (postgres and redis)
+	//////////////////////////////////////////////////////////////////////////
+	log.Infof("Setting up database connections")
+	db, err := dbcommon.OpenPgxPool(config.Postgres)
+	if err != nil {
+		return errors.WithMessage(err, "Error opening connection to postgres")
+	}
+	jobRepository := database.NewPostgresJobRepository(db, int32(config.DatabaseFetchSize))
+	executorRepository := database.NewPostgresExecutorRepository(db)
+
+	redisClient := redis.NewUniversalClient(config.Redis.AsUniversalOptions())
+	defer func() {
+		err := redisClient.Close()
+		if err != nil {
+			log.WithError(errors.WithStack(err)).Warnf("Redis client didn't close down cleanly")
+		}
+	}()
+	queueRepository := database.NewLegacyQueueRepository(redisClient)
+
+	//////////////////////////////////////////////////////////////////////////
+	// Pulsar
+	//////////////////////////////////////////////////////////////////////////
+	log.Infof("Setting up Pulsar connectivity")
+	pulsarClient, err := pulsarutils.NewPulsarClient(&config.Pulsar)
+	defer pulsarClient.Close()
+	if err != nil {
+		return errors.WithMessage(err, "Error creating pulsar client")
+	}
+	pulsarPublisher, err := NewPulsarPublisher(pulsarClient, pulsar.ProducerOptions{
+		Name:             fmt.Sprintf("armada-scheduler-%s", uuid.New()),
+		CompressionType:  config.Pulsar.CompressionType,
+		CompressionLevel: config.Pulsar.CompressionLevel,
+		BatchingMaxSize:  config.Pulsar.MaxAllowedMessageSize,
+		Topic:            config.Pulsar.JobsetEventsTopic,
+	}, config.PulsarSendTimeout)
+	if err != nil {
+		return errors.WithMessage(err, "Error creating pulsar publisher")
+	}
+
+	//////////////////////////////////////////////////////////////////////////
+	// Leader Election
+	//////////////////////////////////////////////////////////////////////////
+	leaderController, err := createLeaderController(config.Leader)
+	if err != nil {
+		return errors.WithMessage(err, "Error creating leader controller")
+	}
+	g.Go(func() error { return leaderController.Run(ctx) })
+
+	//////////////////////////////////////////////////////////////////////////
+	// Executor Api
+	//////////////////////////////////////////////////////////////////////////
+	log.Infof("Setting up executor api")
+	apiProducer, err := pulsarClient.CreateProducer(pulsar.ProducerOptions{
+		Name:             fmt.Sprintf("armada-executor-api-%s", uuid.New()),
+		CompressionType:  config.Pulsar.CompressionType,
+		CompressionLevel: config.Pulsar.CompressionLevel,
+		BatchingMaxSize:  config.Pulsar.MaxAllowedMessageSize,
+		Topic:            config.Pulsar.JobsetEventsTopic,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "Error creating pulsar producer for executor api")
+	}
+	authServices := auth.ConfigureAuth(config.Auth)
+	grpcServer := grpcCommon.CreateGrpcServer(config.Grpc.KeepaliveParams, config.Grpc.KeepaliveEnforcementPolicy, authServices)
+	defer grpcServer.GracefulStop()
```

Comment:
You may want to close forcefully on a a timeout (like we do in `server.go`), since I think `GracefulStop` may hang indefinitely while waiting for connections to close.

Reply:
I think this should be fine- the armada server has to deal with unbounded streams and rpcs that last hours.  This should be dealing with rpcs that last seconds at most.